client/db/bolt/db.go

```diff
@@ -0,0 +1,535 @@
+	orderB := oBkt.Get(orderKey)
+	if orderB == nil {
+		return nil, fmt.Errorf("nil order bytes for order %x", oid)
+	}
+	ord, err := order.DecodeOrder(bCopy(orderB))
+	if err != nil {
+		return nil, fmt.Errorf("error decoding order %x: %v", oid, err)
+	}
+	proofB := oBkt.Get(proofKey)
+	if proofB == nil {
+		return nil, fmt.Errorf("nil proof for order %x", oid)
+	}
+	proof, err := dexdb.DecodeOrderProof(bCopy(proofB))
+	if err != nil {
+		return nil, fmt.Errorf("error decoding order proof for %x: %v", oid, err)
+	}
+	return &dexdb.MetaOrder{
+		MetaData: &dexdb.OrderMetaData{
+			Proof:  *proof,
+			Status: order.OrderStatus(intCoder.Uint16(oBkt.Get(statusKey))),
+			DEX:    string(oBkt.Get(dexKey)),
+		},
+		Order: ord,
+	}, nil
+}
+
+// ordersView is a convenience function for reading from the order bucket.
+func (db *boltDB) ordersView(f bucketFunc) error {
+	return db.withBucket(ordersBucket, db.View, f)
+}
+
+// ordersUpdate is a convenience function for updating the order bucket.
+func (db *boltDB) ordersUpdate(f bucketFunc) error {
+	return db.withBucket(ordersBucket, db.Update, f)
+}
+
+// UpdateMatch updates the match information in the database. Any existing
+// entry for the same match ID will be overwritten without indication.
+func (db *boltDB) UpdateMatch(m *dexdb.MetaMatch) error {
+	match, md := m.Match, m.MetaData
+	if md.Quote == md.Base {
+		return fmt.Errorf("quote and base asset cannot be the same")
+	}
+	if md.DEX == "" {
+		return fmt.Errorf("empty DEX not allowed")
+	}
+	return db.matchesUpdate(func(master *bolt.Bucket) error {
+		mBkt, err := master.CreateBucketIfNotExists(match.MatchID[:])
+		if err != nil {
+			return fmt.Errorf("order bucket error: %v", err)
+		}
+		return newBucketPutter(mBkt).
+			put(baseKey, uint32Bytes(md.Base)).
+			put(quoteKey, uint32Bytes(md.Quote)).
+			put(statusKey, []byte{byte(md.Status)}).
+			put(dexKey, []byte(md.DEX)).
+			put(updateTimeKey, uint64Bytes(timeNow())).
+			put(proofKey, md.Proof.Encode()).
+			put(matchKey, order.EncodeMatch(match)).
+			err()
+	})
+}
+
+// ActiveMatches retrieves the matches that are in an active state, which is
+// any state except order.MatchComplete.
+func (db *boltDB) ActiveMatches() ([]*dexdb.MetaMatch, error) {
+	return db.filteredMatches(func(mBkt *bolt.Bucket) bool {
+		status := mBkt.Get(statusKey)
+		return len(status) != 1 || status[0] != uint8(order.MatchComplete)
+	})
+}
+
+// filteredMatches gets all matches that pass the provided filter function. Each
+// match's bucket is provided to the filter, and a boolean true return value
+// indicates the match should be decoded and returned.
+func (db *boltDB) filteredMatches(filter func(*bolt.Bucket) bool) ([]*dexdb.MetaMatch, error) {
+	var matches []*dexdb.MetaMatch
+	var err error
+	db.matchesView(func(master *bolt.Bucket) error {
+		master.ForEach(func(k, _ []byte) error {
+			mBkt := master.Bucket(k)
+			if mBkt == nil {
+				return fmt.Errorf("match %x bucket is not a bucket", k)
```

Comment:
Pretty sure these errors are still getting tossed out as they come out of `matchesView`.